### DISCUSSION TITLE
Github Actions services not reachable

### DISCUSSION THREAD

**Discussion by juulSme** [community/community#26114] [Actions]
<p>I’m trying to use Github Actions with a Docker compose-like setup. I would like to use a service container to start a postgres instance, to connect to from an elixir container. The way I read the <a href="https://help.github.com/en/articles/workflow-syntax-for-github-actions#jobsjob_idservices" rel="nofollow noopener">docs</a>, this should work:</p>
<pre><code>name: CI / Automated testing
on: push

jobs:
  test:
    name: Unit &amp; integration tests
    runs-on: ubuntu-latest

    services:
      postgres:
        image: postgres:11
        ports:
          - 5432:5432
        env:
          POSTGRES_PASSWORD: supersecret

    container:
      image: elixir:1.9-slim
      env:
        MIX_ENV: test
        POSTGRES_PORT: 5432
        POSTGRES_HOSTNAME: postgres
        POSTGRES_PASSWORD: supersecret

    steps:
      - name: Checkout code
        uses: actions/checkout@master
      - name: Install dependencies
          mix do local.hex --force, local.rebar --force, deps.get
      - name: Compile code (warnings as errors)
        run: mix compile --warnings-as-errors
      - name: Check code formatting
        run: mix format --check-formatted
      - name: Run tests
        run: mix test
</code></pre>
<p>But sadly it does not work. Setting POSTGRES_HOSTNAME to localhost also does not seem to work. Does anybody know how to fix the issue?</p>

**Comment by juulSme** [comment]
<p>And yes, those environment variables are correctly picked up by the application and it really tries to connect to the hostname set in the variable :wink:</p>

**Comment by juulSme** [comment]
<p>Anybody? <a class="mention" href="/jeremyepling">@jeremyepling</a>, might you be able to help? I cannot for the life of me get this working.</p>
<pre><code>name: CI / Automated testing
on: push

jobs:
  test:
    name: Unit &amp; integration tests
    runs-on: ubuntu-latest

    services:
      postgres:
        image: postgres:11
        ports:
          - 5432:5432
        env:
          POSTGRES_PASSWORD: supersecret

    container:
      image: elixir:1.9-slim
      env:
        MIX_ENV: test
        POSTGRES_HOSTNAME: localhost
        POSTGRES_PASSWORD: supersecret

    steps:
      - name: Display build environment
        run: printenv
      - name: Checkout code
        uses: actions/checkout@master
      - name: Install dependencies
        run: mix do local.hex --force, local.rebar --force, deps.get
      - name: Run tests
        run: mix test
        env:
          POSTGRES_PORT: ${{ job.services.postgres.ports['5432'] }}
</code></pre>
<p>I’ve tried:</p>
<p> - setting hostname to “postgres” and “localhost”</p>
<p> - setting service port to 5432 and 5432:5432</p>
<p> - setting env POSTGRES_PORT to 5432 and ${{ job.services.postgres.ports[‘5432’] }}</p>

**Comment by juulSme** [comment]
<p>Or you <a class="mention" href="/ethomson">@ethomson</a> ?</p>

**Comment by dakale** [comment, marked as answer]
<p>One possibility is that the postgres service is not fully started by the time your steps try to access it (I’ve seen this issue for mysql, postgres, etc)</p>
<p>You can try adding a healthcheck via options (or in the Dockerfile, if you were building your own service container). The actions runner will respect the health status of the services before finalizing the “Initialize containers” step:</p>
<pre><code>services:
postgress:
  image: postgres:11
   options: &gt;-
    --health-cmd pg_isready
     --health-interval 10s
     --health-timeout 5s
     --health-retries 5
</code></pre>
<p>Sorry if the formatting got messed up via pasting here, and hope that helps</p>

**Comment by dakale** [comment]
<p>Additionally, just in case I skimmed your question too fast and missed something:</p>
<p>All the containers in a job are connected to the same Docker network. This means you can access other containers using their network alias (we assign the network alias based on the service name), which in your case is “postgres”, so you can also try using “postgres” as the POSTGRES_HOSTNAME inside your steps there. </p>
<p>As a nit, you do not need to expose ports explictly here, because all the containers on Docker network automatically expose ports to each other (<a href="https://docs.docker.com/network/bridge/" rel="nofollow noopener">https://docs.docker.com/network/bridge/</a>) </p>
<p>If you were *not* using a job container, you would be correct in needing to access the service via localhost, and needing to bind a port on the host machine</p>

**Comment by tschersach** [comment]
<p>I have the exact same problem as <a class="mention" href="/juulsme">@juulsme</a> but with my dynamodb-local service</p>
<pre><code>services:
dynamodb:
image: amazon/dynamodb-local
ports:
- 8000:8000
</code></pre>
<p>being unreachable from localhost:8000.</p>
<p>I’m currently trying to inspect the container because a simple <code>docker ps -a</code> during my steps shows that the container is stopped … </p>
<p>EDIT SOLVED:</p>
<p>workdir is overwritten in services with the path to a volume where you repo codes lives.</p>
<p>I think this should be optionnal and easily opted-out because I think the main use case for services is just to run a service that will never interact with your main repo code.</p>
<p>The fix for running dynamodb-local in a service is to switch back to the old workdir where the dynamodb-local JAR is located.</p>
<pre><code>services:
dynamodb:
image: amazon/dynamodb-local
options: --workdir /home/dynamodblocal
ports:
- 8000:8000
</code></pre>
<p>N.B : I’ve found the correct workdir by using <code>docker history</code> on the image as the Dockerfile for dynamodb-local is not public (I think) </p>

**Comment by juulSme** [comment]
<p>Thank you for your response, I will try with the health check tomorrow! (it’s midnight where I live now). I know about the docker networks, I’ve tried using the container hostname first - that seemed most logical to me as well and I also spotted that the containers are started with the same network flag. I’ll give it a go with the healthcheck :slight_smile:</p>

**Comment by juulSme** [comment]
<p>And I’ll try the workdir thing too :slight_smile:</p>

**Comment by juulSme** [comment]
<p>The healthcheck thing did the trick, although I’ll keep the changed workdir in mind for future reference. Thanks for your help!</p>
<p>In the end, the solution was to “do things like in compose” plus the healtcheck:</p>
<pre><code>name: CI / Automated testing
on:
  push:
    branches:

jobs:
  test:
    name: Unit &amp; integration tests
    runs-on: ubuntu-latest

    services:
      postgres:
        image: postgres:11
        env:
          POSTGRES_PASSWORD: supersecret
        options: &gt;-
          --health-cmd pg_isready
          --health-interval 10s
          --health-timeout 5s
          --health-retries 5

    container:
      image: elixir:1.9-slim
      env:
        POSTGRES_HOSTNAME: postgres
        POSTGRES_PASSWORD: supersecret
        POSTGRES_PORT: 5432

    steps:
      - name: Checkout code
        uses: actions/checkout@master
      - name: Install dependencies
        run: mix do local.hex --force, local.rebar --force, deps.get
      - name: Run tests
        run: mix test
</code></pre>

**Reply by FengJi2021** [reply]
I have the same issue but health check not working. Strange is that only one container is running when I have 2 services. So i think maybe some pre config issue with act?
